auth/src/main/java/com/firebase/ui/auth/viewmodel/smartlock/SmartLockHandler.java

```diff
@@ -51,44 +61,54 @@ public void onActivityResult(int requestCode, int resultCode) {
-        saveCredentials(CredentialsUtil.buildCredential(firebaseUser, password, accountType));
+        saveCredentials(CredentialsUtils.buildCredential(firebaseUser, password, accountType));
     }
 
     /** Initialize saving a credential. */
-    public void saveCredentials(Credential credential) {
+    public void saveCredentials(@Nullable Credential credential) {
         if (!getArguments().enableCredentials) {
-            setResult(Resource.forVoidSuccess());
+            setResult(Resource.forSuccess(mResponse));
             return;
         }
-        setResult(Resource.<Void>forLoading());
+        setResult(Resource.<IdpResponse>forLoading());
 
         if (credential == null) {
-            setResult(Resource.<Void>forFailure(new FirebaseUiException(ErrorCodes.UNKNOWN_ERROR,
-                    "Failed to build credential.")));
+            setResult(Resource.<IdpResponse>forFailure(new FirebaseUiException(
+                    ErrorCodes.UNKNOWN_ERROR, "Failed to build credential.")));
             return;
         }
 
+        deleteUnusedCredentials();
         getCredentialsClient().save(credential)
                 .addOnCompleteListener(new OnCompleteListener<Void>() {
                     @Override
                     public void onComplete(@NonNull Task<Void> task) {
                         if (task.isSuccessful()) {
-                            setResult(Resource.forVoidSuccess());
+                            setResult(Resource.forSuccess(mResponse));
                         } else if (task.getException() instanceof ResolvableApiException) {
                             ResolvableApiException rae = (ResolvableApiException) task.getException();
-                            setResult(Resource.<Void>forFailure(
+                            setResult(Resource.<IdpResponse>forFailure(
                                     new PendingIntentRequiredException(
                                             rae.getResolution(), RequestCodes.CRED_SAVE)));
                         } else {
                             Log.w(TAG, "Non-resolvable exception: " + task.getException());
-
-                            FirebaseUiException exception = new FirebaseUiException(
+                            setResult(Resource.<IdpResponse>forFailure(new FirebaseUiException(
                                     ErrorCodes.UNKNOWN_ERROR,
                                     "Error when saving credential.",
-                                    task.getException());
-                            setResult(Resource.<Void>forFailure(exception));
+                                    task.getException())));
                         }
                     }
                 });
     }
+
+    private void deleteUnusedCredentials() {
+        if (mResponse.getProviderType().equals(GoogleAuthProvider.PROVIDER_ID)) {
+            // Since Google accounts upgrade email ones, we don't want to end up
```

Comment:
Ok so want to make sure what's going on here:
Whenever we sign in as a Google user we look for matching email credentials and delete them since the Google one is "superior"?

Reply:
Close. You should try this yourself: the "bug" everyone has been complaining about is that when you use the Google provider, it basically overwrites existing providers and pretends that they never existed. AFAIK, this is expected behavior so this removes  the no-longer valid credential. Hope that clarifies things! 😄